src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CryptographicOperations.cs

```diff
@@ -58,5 +64,766 @@ public static void ZeroMemory(Span<byte> buffer)
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="hashAlgorithm"/> has a <see cref="HashAlgorithmName.Name" /> that is
+        ///   <see langword="null" />.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="hashAlgorithm"/> has a <see cref="HashAlgorithmName.Name" /> that is empty.
+        /// </exception>
+        /// <exception cref="PlatformNotSupportedException">
+        ///   <paramref name="hashAlgorithm"/> specifies a hash algorithm not supported by the current platform.
+        /// </exception>
+        /// <exception cref="CryptographicException">
+        ///   <paramref name="hashAlgorithm"/> specifies an unknown hash algorithm.
+        /// </exception>
+        public static byte[] HashData(HashAlgorithmName hashAlgorithm, ReadOnlySpan<byte> source)
+        {
+            int hashSizeInBytes = CheckHashAndGetLength(hashAlgorithm);
+            Debug.Assert(hashAlgorithm.Name is not null);
+
+            byte[] buffer = new byte[hashSizeInBytes];
+            int written = HashProviderDispenser.OneShotHashProvider.HashData(hashAlgorithm.Name, source, buffer);
+            Debug.Assert(written == hashSizeInBytes);
+            return buffer;
+        }
+
+        /// <summary>
+        /// Computes the hash of data.
+        /// </summary>
+        /// <param name="hashAlgorithm">The algorithm used to compute the hash.</param>
+        /// <param name="source">The data to hash.</param>
+        /// <param name="destination">The buffer to receive the hash value.</param>
+        /// <returns>The total number of bytes written to <paramref name="destination" />.</returns>
+        /// <exception cref="ArgumentException">
+        ///   <para>The buffer in <paramref name="destination"/> is too small to hold the calculated hash size.</para>
+        ///   <para>-or-</para>
+        ///   <para><paramref name="hashAlgorithm"/> has a <see cref="HashAlgorithmName.Name" /> that is empty.</para>
+        /// </exception>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="hashAlgorithm"/> has a <see cref="HashAlgorithmName.Name" /> that is
+        ///   <see langword="null" />.
+        /// </exception>
+        /// <exception cref="PlatformNotSupportedException">
+        ///   <paramref name="hashAlgorithm"/> specifies a hash algorithm not supported by the current platform.
+        /// </exception>
+        /// <exception cref="CryptographicException">
+        ///   <paramref name="hashAlgorithm"/> specifies an unknown hash algorithm.
+        /// </exception>
+        public static int HashData(HashAlgorithmName hashAlgorithm, ReadOnlySpan<byte> source, Span<byte> destination)
+        {
+            if (!TryHashData(hashAlgorithm, source, destination, out int bytesWritten))
+            {
+                throw new ArgumentException(SR.Argument_DestinationTooShort, nameof(destination));
```

Comment:
This shows up at least 6 times. Might be worth creating a private static throw helper method for it, both to reduce the duplication and increase the liklihood this will be inlined.

Reply:
There are some subtle differences between these. Some are on ROS, some are on ROM. That would mean using `.Span` on the memory, which can be a little expensive. Or we have two helpers, which I am not sure helps at that point.

Unless you meant for the helper to be

```C#
private static void CheckDestinationSize(int requiredSize, int destinationSize)
{
    if (destinationSize < requiredSize)
    {
        throw new ArgumentException(SR.Argument_DestinationTooShort, "destination");
    }
}

// Used like

CheckDestinationSize(hashLength, destination.Length);
```

Reply:
I actually just meant:
```C#
private static void ThrowDestinationTooShort() => throw new ArgumentException(SR.Argument_DestinationTooShort, "destination");
```
but we could do what you suggest, too.